dbms/src/DataStreams/MultiplexInputStream.h

```diff
@@ -0,0 +1,257 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <DataStreams/IProfilingBlockInputStream.h>
+
+#include <memory>
+#include <queue>
+#include <vector>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+} // namespace ErrorCodes
+
+class MultiPartitionStreamPool
+{
+public:
+    MultiPartitionStreamPool() = default;
+
+    void addPartitionStreams(const BlockInputStreams & cur_streams)
+    {
+        std::unique_lock lk(mu);
+        if (cur_streams.empty())
+            return;
+        streams_queue_by_partition.push_back(
+            std::make_shared<std::queue<std::shared_ptr<IBlockInputStream>>>());
+        for (const auto & stream : cur_streams)
+            streams_queue_by_partition.back()->push(stream);
+        added_streams.insert(added_streams.end(), cur_streams.begin(), cur_streams.end());
+    }
+
+    std::shared_ptr<IBlockInputStream> pickOne()
+    {
+        std::unique_lock lk(mu);
+        if (streams_queue_by_partition.empty())
+            return nullptr;
+        if (streams_queue_id >= static_cast<int>(streams_queue_by_partition.size()))
+            streams_queue_id = 0;
+
+        auto & q_ptr = streams_queue_by_partition[streams_queue_id];
+        auto & q = *q_ptr;
+        std::shared_ptr<IBlockInputStream> ret = nullptr;
+        assert(!q.empty());
+        ret = q.front();
+        q.pop();
+        if (q.empty())
+            streams_queue_id = removeQueue(streams_queue_id);
+        else
+            streams_queue_id = nextQueueId(streams_queue_id);
+        return ret;
+    }
+
+    int exportAddedStreams(BlockInputStreams & ret_streams)
+    {
+        std::unique_lock lk(mu);
+        for (auto & stream : added_streams)
+            ret_streams.push_back(stream);
+        return added_streams.size();
+    }
+
+private:
+    int removeQueue(int queue_id)
+    {
+        streams_queue_by_partition[queue_id] = nullptr;
+        if (queue_id != static_cast<int>(streams_queue_by_partition.size()) - 1)
+        {
+            swap(streams_queue_by_partition[queue_id], streams_queue_by_partition.back());
+        }
+        streams_queue_by_partition.pop_back();
+        if (queue_id < static_cast<int>(streams_queue_by_partition.size()))
```

Comment:
These two if statments seem can be merged into one.

Reply:
this size of streams_queue_by_partition changed after `pop_back`() in L85

Reply:
Think the queue size is 10, when queue id is 9, then first if returns false, seconds returns false; when queue id is less than 9, first if returns true, second if returns true, thus seems works.

Reply:
could you give me the example changed code？

Reply:
if (queue_id != static_cast<int>(streams_queue_by_partition.size()) - 1)
{
swap(streams_queue_by_partition[queue_id], streams_queue_by_partition.back());
streams_queue_by_partition.pop_back();
return queue_id;
}
else
{
streams_queue_by_partition.pop_back();
return 0;
}

Reply:
IMO there is no obvious difference...

Reply:
Fewer if, more efficient, easier to read.

Reply:
since the operation level is file（DMsegMent not row ， the CPU overhead can be ignored.

Reply:
just a personal opinion，the previous one is easier to understand for some people such as me...

Reply:
Yes, you're right, I agree the partition parrel performance issue fixed here reduced much more time than the if statement which executes not frequently. 
Ehh, think this way: a little extra effort to make code more readable and efficient, it would be a nice practice:)

Reply:
ok

Reply:
done